[WIP] enhances persistent connections to use rpc message passing

privateip opened the pull request:
* network_cli now implements connection loca
* network_cli now is a connection plugin
* adds executor/process/connection.py:ConnectionProcess
* moves do_fork() to executor/process/__init__.py
* now checks for instance variable on connection `force_persistent_connection`
* adds utils/jsonrpc.py:JsonRpcServer
* refactors module_utils/connection.py:Connection to accept only socket_path

Commit:
[WIP] enhances persistent connections to use rpc message passing

* network_cli now implements connection loca
* network_cli now is a connection plugin
* adds executor/process/connection.py:ConnectionProcess
* moves do_fork() to executor/process/__init__.py
* now checks for instance variable on connection `force_persistent_connection`
* adds utils/jsonrpc.py:JsonRpcServer
* refactors module_utils/connection.py:Connection to accept only socket_path

bcoca commented on lib/ansible/executor/process/__init__.py:
```diff
@@ -1,20 +1,45 @@
+import os
+import sys
+
+def do_fork():
```
mostly same as https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/service.py#L94

seems like this is a simpler case

Commit:
fix up santiy test errors

Commit:
refactor do_fork into module_utils/service.py and combine with daemonize

jimi-c commented on lib/ansible/module_utils/service.py:
```diff
@@ -149,6 +140,38 @@ def daemonize(module, cmd):
+    return pid
+
+
+def daemonize(module, cmd):
```
Why are we doing all this extra boilerplate when multiprocessing.Process does it for us?

bcoca replied:
this is mostly a copy of preexisting code, we should migrate to multiprocessing now that 2.6 is min target. But my original point is still valid, we should have 1 copy of this code.

ganeshrn commented on lib/ansible/executor/process/connection.py:
```diff
@@ -0,0 +1,204 @@
+        play_context = self.connection._play_context
+
+        ssh = connection_loader.get('ssh', class_only=True)
+        cp = ssh._create_control_path(play_context.remote_addr, play_context.port, play_context.remote_user)
```
As some network os support multiple connections type eg. junos support both network_cli and netconf, I think connection name should also be passed to _create_control_path()

ganeshrn commented on lib/ansible/executor/process/connection.py:
```diff
@@ -0,0 +1,204 @@
+            timeout -= 1
+        else:
+            raise AnsibleError('timeout waiting for local socket', play_context.remote_addr)
+
```
To handle privilege escalation related activities `update_play_context()` should be called somewhere in this function?

ganeshrn commented on lib/ansible/plugins/connection/network_cli.py:
```diff
@@ -68,31 +71,56 @@
+                raise AttributeError("'%s' object has no attribute '%s'" % (self.__class__.__name__, name))
+            return getattr(self._cliconf, name)
+
+    def exec_command(self, cmd, in_data=None, sudoable=True):
```
Should this handle `prompt()`  command to ensure backward compatibility with existing action plugins?

privateip replied:
this should now be fixed in e4eeebd

Commit:
fixes backwards compatibility with current persistent connection framework

removes bin/ansible-connection as it is no longer needed

module_utils/connection now passes rpc calls instead of native commands

privateip commented:
closing this ... superseded by #31417
